plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractIcebergTableOperations.java

```diff
@@ -226,13 +229,25 @@ protected void refreshFromMetadataLocation(String newLocation)
             return;
         }
 
-        TableMetadata newMetadata = Failsafe.with(RetryPolicy.builder()
-                        .withMaxRetries(20)
-                        .withBackoff(100, 5000, MILLIS, 4.0)
-                        .withMaxDuration(Duration.ofMinutes(10))
-                        .abortOn(org.apache.iceberg.exceptions.NotFoundException.class)
-                        .build()) // qualified name, as this is NOT the io.trino.spi.connector.NotFoundException
-                .get(() -> TableMetadataParser.read(fileIo, io().newInputFile(newLocation)));
+        TableMetadata newMetadata;
+        try {
+            newMetadata = Failsafe.with(RetryPolicy.builder()
+                            .withMaxRetries(20)
+                            .withBackoff(100, 5000, MILLIS, 4.0)
+                            .withMaxDuration(Duration.ofMinutes(10))
+                            .abortOn(org.apache.iceberg.exceptions.NotFoundException.class)
+                            .build()) // qualified name, as this is NOT the io.trino.spi.connector.NotFoundException
+                    .get(() -> TableMetadataParser.read(fileIo, io().newInputFile(newLocation)));
+        }
+        catch (RuntimeException e) {
+            if (e.getCause() != null) {
+                if (e.getCause() instanceof FileNotFoundException
+                        || e.getCause().getMessage().contains("The specified key does not exist")) {
```

Comment:
The check against an error message is not ideal.
For S3, look into `TrinoS3FileSystem` whether it would be appropriate to throw a `FileNotFoundException` in such cases 

`io.trino.plugin.hive.s3.TrinoS3FileSystem.TrinoS3InputStream#read(long, byte[], int, int)`

Reply:
what was the exception previously raised here?